### PR TITLE
Improve Animation docs

### PR DESCRIPTION
**Pull request by AntonBallmaier**
## Explanation for Changes

[Documentation Reference](https://manimce--1521.org.readthedocs.build/en/1521/reference/manim.animation.animation.Animation.html)

## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [x] My new functions/classes either have a docstring or are private
- [x] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

**Review by Nikhil0504 — CHANGES_REQUESTED** [review]
Can you also write descriptions for `get_run_time` And `get_run_time`?

**Comment by AntonBallmaier** [comment]
> Can you also write descriptions for `get_run_time` And `get_run_time`?

I guess you meant _and `set_run_time`_. I added docs to all getters and setters.

File: manim/animation/animation.py [hunk 1]
```diff
@@ -22,6 +22,68 @@
...
+
+    Animations have a fixed time span.
+
+    Parameters
+    ----------
+    mobject
+        The mobject to be animated. This is not required for all types of animations.
+    lag_ratio
+        Defines the delay after which the animation is applied to submobjects. This lag
+        is relative to the duration of the animation.
```
Comment (kolibril13) [inline manim/animation/animation.py:44]: Already really nice!
Reading this, I don't completely get, what the word "ratio" means in the context. Means the ratio of 1, that there is no overlap of the submobjects?
And something like "ratio" =    "overlap of two subanimation"  /  "length of one subanimation" ?
I know this formula is wrong, but I did not have the time right now to properly think about it ;)

File: manim/animation/animation.py [hunk 2]
```diff
@@ -22,6 +22,68 @@
...
+    --------
+
+    .. manim:: LagRatios
+
+        class LagRatios(Scene):
+            def construct(self):
+                ratios = [0, 0.1, 0.5, 1, 2]  # demonstrated lag_ratios
+
+                # Create dot groups
+                group = VGroup(*[Dot() for _ in range(4)]).arrange_submobjects()
```
Comment (kolibril13) [inline manim/animation/animation.py:76]:
```suggestion
                group = VGroup()
                for _ in range(0,4):
                    group.add(Dot())
                group.arrange(RIGHT)
```
Zen of python:
Sparse is better than dense.
Readability counts.

File: manim/animation/animation.py [hunk 3]
```diff
@@ -22,6 +22,68 @@
...
+                groups = VGroup(*[group.copy() for _ in ratios]).arrange_submobjects(buff=1)
```
Comment (kolibril13) [inline manim/animation/animation.py:77]: Same here

File: manim/animation/animation.py (hunk 2 above)
Comment (kolibril13) [reply manim/animation/animation.py:76]: In my projects, I also really like to use this long and dense code lines:
VGroup(*[Dot() for _ in range(4)]).arrange_submobjects().
But I think for the docs, shorter and more lines is better.
However, this is free to be discussed a bit more.

Comment (AntonBallmaier) [reply manim/animation/animation.py:76]: If this was a tutorial example on how to use manim I would agree. But in the examples I don't want to have most of the code dedicated to setup unrelated to the feature that should be demonstrated. I also added a comment explaining what the code does. Imo that should be enough.

Comment (kolibril13) [reply manim/animation/animation.py:76]: okay, then I am fine with that!

File: manim/animation/animation.py (hunk 1 above)
Comment (AntonBallmaier) [reply manim/animation/animation.py:44]: I can see where you're coming from - I was struggling to write that description and was never quite happy with it. My approach to solve that problem was to add an example. Imo the combination explains it reasonably well without having a too wordy explanation.

Comment (kolibril13) [reply manim/animation/animation.py:44]: Ture.
There can definitely be fine-tuning, but much better than nothing :)

File: manim/animation/animation.py [hunk 4]
```diff
@@ -192,29 +299,92 @@ def get_sub_alpha(self, alpha: float, index: int, num_submobjects: int) -> float
...
         return self
 
     def get_rate_func(
         self,
     ) -> Callable[[float], float]:
+        """Get the rate function of the animation.
+
+        Returns
+        -------
+        Callable[[float], float]
```
Comment (kolibril13) [inline manim/animation/animation.py:359]:
```suggestion
```
These lines are not necessary, as this is already defined by the typehint -> Callable[[float], float]:

File: manim/animation/animation.py [hunk 5]
```diff
@@ -192,29 +299,92 @@ def get_sub_alpha(self, alpha: float, index: int, num_submobjects: int) -> float
...
+        """Set the name of the animation.
+
+        Parameters
+        ----------
+        name
+            The new name of the animation.
+
+        Returns
+        -------
+        ``Animation``
```
Comment (kolibril13) [inline manim/animation/animation.py:374]: same here

File: manim/animation/animation.py [hunk 6]
```diff
@@ -192,29 +299,92 @@ def get_sub_alpha(self, alpha: float, index: int, num_submobjects: int) -> float
...
+        """
         self.name = name
         return self
 
     def is_remover(self) -> bool:
+        """Test if a the animation is a remover.
+
+        Returns
+        -------
+        ``bool``
```
Comment (kolibril13) [inline manim/animation/animation.py:385]: and here

File: manim/animation/animation.py (hunk 4 above)
Comment (AntonBallmaier) [reply manim/animation/animation.py:360]: I tested it and it actually does not work without it I'm afraid.

Comment (kolibril13) [reply manim/animation/animation.py:360]: okay, I did not know that.
But then, I am also not sure why we have this typehint at all.

File: manim/animation/animation.py [hunk 7]
```diff
@@ -22,6 +31,68 @@
...
+        is relative to the duration of the animation.
+
+        This does not influence the total
+        runtime of the animation. Instead the runtime of individual animations is
+        adjusted so that the complete animation has the defined run time.
+
+    run_time
+        The duration of the animation in seconds.
+    rate_func
+        The function defining the animation progress based on the relative runtime.
```
Comment (kolibril13) [inline manim/animation/animation.py:53]:
```suggestion
        The function defining the animation progress based on the relative runtime (see  :mod:`~.rate_functions`) .
```

File: manim/animation/animation.py [hunk 8]
```diff
@@ -22,6 +31,68 @@
...
+
+        For example ``rate_func(0.5)`` is the proportion of the animation that is done
+        after half of the animations run time.
+    name
+        The name of the animation. This gets displayed while rendering the animation.
+        Defualts to <class-name>(<Mobject-name>).
```
Comment (kolibril13) [inline manim/animation/animation.py:59]:
```suggestion
        Defaults to <class-name>(<Mobject-name>).
```

File: manim/animation/animation.py (hunk 4 above)
Comment (AntonBallmaier) [reply manim/animation/animation.py:360]: For Intellisense I guess :man_shrugging:

**Review by kolibril13 — APPROVED** [review]

File: manim/animation/animation.py (hunk 4 above)
Comment (kolibril13) [reply manim/animation/animation.py:360]: Ah,  makes sense

**Review by Nikhil0504 — APPROVED** [review]
LGTM

File: manim/animation/animation.py [hunk 9]
```diff
@@ -112,21 +181,45 @@ def begin(self) -> None:
...
         if self.is_remover():
             scene.remove(self.mobject)
 
     def create_starting_mobject(self) -> Mobject:
         # Keep track of where the mobject starts
         return self.mobject.copy()
 
-    def get_all_mobjects(self) -> Tuple[Mobject, Mobject]:
-        """
+    def get_all_mobjects(self) -> Sequence[Mobject]:
```
Comment (naveen521kk) [inline manim/animation/animation.py:214]: I think the previous typing gives more sense since it is always going to return a tuple where there are two Mobject.

File: manim/animation/animation.py [hunk 10]
```diff
@@ -192,29 +300,92 @@ def get_sub_alpha(self, alpha: float, index: int, num_submobjects: int) -> float
...
+        """Set the name of the animation.
+
+        Parameters
+        ----------
+        name
+            The new name of the animation.
+
+        Returns
+        -------
+        ``Animation``
```
Comment (naveen521kk) [inline manim/animation/animation.py:375]: use 
```suggestion
        :class:`Animation`
```

Comment (naveen521kk) [reply manim/animation/animation.py:375]: this needs to be changed everywhere

File: manim/animation/animation.py (hunk 9 above)
Comment (AntonBallmaier) [reply manim/animation/animation.py:214]: The reason I changed it, is that subclasses often return more than these two mobjects. Therefor I think Sequence[Mobject] is a better typehint, since it fits to the method in general and not only to the implementation in Animation. However, I don't have a strong opinion about this.

File: manim/animation/animation.py (hunk 10 above)
Comment (AntonBallmaier) [reply manim/animation/animation.py:375]: I feel like this is a stylistic choice we should discuss in general - do we really want to link to the page we are already on in every other sentence? I find it a bit unnecessary and cluttered to do that...

File: manim/animation/animation.py (hunk 9 above)
Comment (naveen521kk) [reply manim/animation/animation.py:215]: Makes sense. But still, I could see https://github.com/ManimCommunity/manim/search?q=get_all_mobjects that each subclass returning a different thing, like some returns a Group and other a List. So, I think it is better to have `Tuple[Mobject, Mobject]` unless we change all the inherited subclass to return `Sequence[Mobject]`.

File: manim/animation/animation.py (hunk 10 above)
Comment (naveen521kk) [reply manim/animation/animation.py:408]: Or you can remove the ` there. https://numpydoc.readthedocs.io/en/latest/format.html#returns
```suggestion
        Animation
```

Comment (AntonBallmaier) [reply manim/animation/animation.py]: > Or you can remove the ` there

I did that :)

File: manim/animation/animation.py (hunk 9 above)
Comment (AntonBallmaier) [reply manim/animation/animation.py:215]: I did the latter :)

File: manim/animation/composition.py [hunk 11]
```diff
@@ -46,7 +46,7 @@ def __init__(
         super().__init__(self.group, rate_func=rate_func, lag_ratio=lag_ratio, **kwargs)
         self.run_time: float = self.init_run_time(run_time)
 
-    def get_all_mobjects(self) -> Group:
+    def get_all_mobjects(self) -> Sequence[Mobject]:
```
Comment (naveen521kk) [inline manim/animation/composition.py:49]: hmm, it's an incompatible type, right?

Comment (AntonBallmaier) [reply manim/animation/composition.py:49]: is it? You can Iterate it and use group[index] - isn't that all a Sequence needs?

Comment (AntonBallmaier) [reply manim/animation/composition.py:49]: Anyway, now it should be fitting :)

File: manim/animation/composition.py [hunk 12]
```diff
@@ -46,8 +46,8 @@ def __init__(
         super().__init__(self.group, rate_func=rate_func, lag_ratio=lag_ratio, **kwargs)
         self.run_time: float = self.init_run_time(run_time)
 
-    def get_all_mobjects(self) -> Group:
-        return self.group
+    def get_all_mobjects(self) -> Sequence[Mobject]:
+        return list(self.group)
```
Comment (naveen521kk) [inline manim/animation/composition.py:50]: isn't this a breaking change? maybe add it to changelog?

Comment (AntonBallmaier) [reply manim/animation/composition.py:50]: You are right, I'll add that. However, I don't think anyone used that that method returned a group anywhere as it is a quite low level method.
